analyzers/src/SonarAnalyzer.Common/Rules/FindInsteadOfFirstOrDefaultBase.cs

```diff
@@ -0,0 +1,64 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Rules;
+
+public abstract class FindInsteadOfFirstOrDefaultBase<TSyntaxKind, TInvocationExpression> : SonarDiagnosticAnalyzer<TSyntaxKind>
+    where TSyntaxKind : struct
+    where TInvocationExpression : SyntaxNode
+{
+    private const string DiagnosticId = "S6602";
+
+    protected override string MessageFormat => $"\"{nameof(Array.Find)}\" method should be used instead of the \"{nameof(Enumerable.FirstOrDefault)}\" extension method.";
+
+    private static readonly ImmutableArray<KnownType> RuleSpecificTypes = ImmutableArray.Create(
+        KnownType.System_Collections_Generic_List_T,
+        KnownType.System_Array,
+        KnownType.System_Collections_Immutable_ImmutableList_T);
+
+    protected abstract bool HasOneArgument(TInvocationExpression invocation);
```

Comment:
In [this PR](https://github.com/SonarSource/sonar-dotnet/pull/7208/files#diff-16a2d259d04a1b16e0c004f5d441adaf8e932260e8a07facf9e70e9d4bd123fc) I'm adding the HasExactlyNArguments method in the Facade, I think you can copy that implementation, this way you'll be able to put all the checks in the baseclass